torch/_inductor/fx_passes/b2b_gemm.py

```diff
@@ -154,118 +298,339 @@ def cdiv(x: int, y: int):
+    subgraph: Subgraph,
+    A: torch.Tensor,
+    B: torch.Tensor,
+    C: torch.Tensor,
+    *,
+    out: torch.Tensor,
+) -> torch.Tensor:
+    if is_left_assoc:
+        torch.mm(subgraph.graph_module(torch.mm(A, B)), C, out=out)
+    else:
+        torch.mm(A, subgraph.graph_module(torch.mm(B, C)), out=out)
     return out
 
 
 unoptimized_choice = ExternKernelChoice(unoptimized_b2b_gemm)
 
 
 def tuned_b2b_gemm(
-    mat1: torch._inductor.ir.TensorBox,
-    mat2: torch._inductor.ir.TensorBox,
-    mat3: torch._inductor.ir.TensorBox,
+    is_left_assoc: bool,
+    subgraph: Subgraph,
+    A: torch._inductor.ir.TensorBox,
+    B: torch._inductor.ir.TensorBox,
+    C: torch._inductor.ir.TensorBox,
     *,
     layout=None,
 ) -> torch._inductor.ir.TensorBox:
     # call .realize() to get rid of Pointwise
-    mat1.realize()
-    mat2.realize()
-    mat3.realize()
+    A.realize()
+    B.realize()
+    C.realize()
     layout = FixedLayout(
-        mat1.get_device(), mat1.get_dtype(), [mat1.shape[0], mat3.shape[1]]
+        A.get_device(), A.get_dtype(), [A.shape[0], C.shape[1]]
+    )
+    subgraph_buffer = build_subgraph_buffer(
+        [create_placeholder("inner_mm", A.get_dtype(), A.get_device())],
+        subgraph,
     )
     choices: list[TritonTemplateCaller] = []
     for config in b2b_gemm_configs:
-        b2b_gemm_template.maybe_append_choice(
-            choices,
-            input_nodes=(mat1, mat2, mat3),
-            layout=layout,
-            **config,
-        )
+        if is_left_assoc:
+            b2b_gemm_left_template.maybe_append_choice(
+                choices,
+                input_nodes=(A, B, C),
+                layout=layout,
+                subgraphs=[subgraph_buffer],
+                **config,
+            )
+        else:
+            b2b_gemm_right_template.maybe_append_choice(
+                choices,
+                input_nodes=(A, B, C),
+                layout=layout,
+                subgraphs=[subgraph_buffer],
+                **config,
+            )
     # add the unoptimized choice to mitigate performance degradation
-    choices.append(unoptimized_choice.bind((mat1, mat2, mat3), layout))
-    return autotune_select_algorithm("b2b_gemm", choices, [mat1, mat2, mat3], layout)
+    choices.append(
+        unoptimized_choice.bind((A, B, C), layout, is_left_assoc=is_left_assoc, subgraph=subgraph)
+    )
+    # autotune
+    return autotune_select_algorithm("b2b_gemm", choices, [A, B, C], layout)
 
 
-# currently it matches ((A @ B) @ C)
-# TODO: later will change to matching (A @ B) in (epilogue2 ((epilogue1 (A @ B)) @ C)) and inspecting the graph
-# TODO: match more cases such as bmm and addmm, and (A @ (B @ C)), etc.
+# match the inner mm of a potential b2b_gemm
 @register_graph_pattern(
-    CallFunction(
-        torch.ops.aten.mm, CallFunction(torch.ops.aten.mm, Arg(), Arg()), Arg()
-    ),
-    extra_check=can_apply_b2b_gemm,
+    CallFunction(torch.ops.aten.mm, Arg(), Arg()),
```

Comment:
The change to support addmm here shouldnt be too hard - see how it's supported here. https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/mm.py#L325C29-L325C43 

Could even just unfuse the bias add